### PR TITLE
fix(react-button): apply hover styles only on devices that support it

### LINKED ISSUE #30388: [Bug]: Sticky hover effect on mobile
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
System:
    OS: macOS 14.2.1
    CPU: (8) arm64 Apple M1
    Memory: 63.02 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Browsers:
    Safari: 17.2.1
    
NPM packages:
"@fluentui/react-components": "^9.45.0"
```

### Are you reporting Accessibility issue?

no

### Reproduction

https://codesandbox.io/p/sandbox/fluent-sticky-hover-cxcx89

## Actual Behavior
In Safari mobile or Chromium mobile, after touching a button, it remains being shown in the "hover" styles. 

image

[Edit @behowell] This also repros on Edge on Windows when using touch.

## Expected Behavior
In mobile browsers, :hover styles should be disabled for all components. There's no obvious way to do this for Fluent v9 components right now.

### Logs

_No response_

### Requested priority

Normal

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by mainframev**
We have a "sticky hover" issue on pointer devices due to the lack of differentiation between pointer and hover devices. This problem also exists in v8.

I approached it with the help of [media-query any-hover](https://www.w3.org/TR/mediaqueries-4/#any-input). Hover effect is wrapped into that query, left `hover:active` outside, so there will be a visual indication after tap on mobile devices. 

### Browser support

[Support is solid](https://caniuse.com/?search=media%20any-hover)

### Devices
While the any-hover media query generally helps distinguish between pointer and hover devices, we must be aware of devices with multiple input methods (e.g., pointer devices connected to keyboards and mouse, or laptops with touchscreens). The `any-hover: hover` query targets any available input method capable of hover, whereas `hover: hover` applies only to the primary input.

### Previous similar issues I found

-  The problem was with Chromium, but the pointer/hover topic was touched. [Mobile: Avoid sticky hover scenarios #5734](https://github.com/microsoft/fluentui/issues/5734)
  - [Closed PR](https://github.com/microsoft/fluentui/pull/31700)

## Previous Behavior
https://github.com/microsoft/fluentui/assets/14054752/e53e7dfd-0888-49ae-b69b-058e09eca655

## New Behavior
https://github.com/microsoft/fluentui/assets/14054752/076f358c-864e-415b-9365-0ebd511090cf

https://github.com/microsoft/fluentui/assets/14054752/19d4f492-4b37-4a91-a524-d7b5cd514356

## Related Issue(s)

- Fixes #30388

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/Avatar-mount_base.html">664</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/Avatar-mount.html">623</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/Button-mount_base.html">311</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/Button-mount.html">312</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/Field-mount_base.html">1149</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/Field-mount.html">1131</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/FluentProvider-mount_base.html">719</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/FluentProvider-mount.html">714</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/FluentProviderWithTheme-mount_base.html">88</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/FluentProviderWithTheme-mount.html">83</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">35</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">42</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">76</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">73</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/MakeStyles-mount_base.html">863</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/MakeStyles-mount.html">869</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/Persona-mount_base.html">1826</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/Persona-mount.html">1724</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/SpinButton-mount_base.html">1430</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/SpinButton-mount.html">1375</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/SwatchPicker-mount_base.html">1567</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31700/perf-test-react-components/SwatchPicker-mount.html">1553</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports | Baseline (minified/GZIP) | PR    | Change     |
| :---------------- | -----------------------: | ----: | ---------: |
| <samp>react-breadcrumb</samp> <br /> <abbr title='bundle-size/index.fixture.js'>@fluentui/react-breadcrumb - package</abbr> | `114.488 kB`<br />`31.824 kB` | `116.704 kB`<br />`32.032 kB` | `2.216 kB` <br />`208 B` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/Button.fixture.js'>Button</abbr> | `37.105 kB`<br />`10.785 kB` | `39.327 kB`<br />`10.909 kB` | `2.222 kB` <br />`124 B` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/CompoundButton.fixture.js'>CompoundButton</abbr> | `43.516 kB`<br />`12.076 kB` | `46.09 kB`<br />`12.25 kB` | `2.574 kB` <br />`174 B` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/MenuButton.fixture.js'>MenuButton</abbr> | `41.886 kB`<br />`12.133 kB` | `44.149 kB`<br />`12.314 kB` | `2.263 kB` <br />`181 B` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/SplitButton.fixture.js'>SplitButton</abbr> | `49.898 kB`<br />`13.725 kB` | `52.418 kB`<br />`13.921 kB` | `2.52 kB` <br />`196 B` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/ToggleButton.fixture.js'>ToggleButton</abbr> | `53.032 kB`<br />`12.545 kB` | `56.717 kB`<br />`12.735 kB` | `3.685 kB` <br />`190 B` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr> | `69.141 kB`<br />`20.157 kB` | `71.361 kB`<br />`20.338 kB` | `2.22 kB` <br />`181 B` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `212.636 kB`<br />`61.23 kB` | `214.852 kB`<br />`61.435 kB` | `2.216 kB` <br />`205 B` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ReactComponents.fixture.js'>react-components: entire library</abbr> | `1.095 MB`<br />`270.838 kB` | `1.099 MB`<br />`271.271 kB` | `4.326 kB` <br />`433 B` |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports | Size (minified/GZIP) |
| ----------------- | -------------------: |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/Card.All.fixture.js'>Card - All</abbr> | `101.437 kB`<br />`29.153 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/Card.fixture.js'>Card</abbr> | `94.22 kB`<br />`27.336 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardFooter.fixture.js'>CardFooter</abbr> | `14.356 kB`<br />`5.798 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardHeader.fixture.js'>CardHeader</abbr> | `16.879 kB`<br />`6.664 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardPreview.fixture.js'>CardPreview</abbr> | `14.42 kB`<br />`5.934 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr> | `44.442 kB`<br />`14.607 kB` |
| <samp>react-dialog</samp> <br /> <abbr title='bundle-size/Dialog.fixture.js'>Dialog (including children components)</abbr> | `99.433 kB`<br />`30.054 kB` |
| <samp>react-message-bar</samp> <br /> <abbr title='bundle-size/Default.fixture.js'>MessageBar (all components)</abbr> | `24.413 kB`<br />`9.117 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr> | `8.39 kB`<br />`2.64 kB` |
| <samp>react-tag-picker</samp> <br /> <abbr title='bundle-size/Package.fixture.js'>@fluentui/react-tag-picker - package</abbr> | `181.063 kB`<br />`54.549 kB` |
| <samp>react-timepicker-compat</samp> <br /> <abbr title='bundle-size/TimePicker.fixture.js'>TimePicker</abbr> | `103.487 kB`<br />`34.632 kB` |
</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/01e9388d19beb2b0c7438ee4a4b1f734b7fe46cd'>01e9388d19beb2b0c7438ee4a4b1f734b7fe46cd</a></sub>
<div id="bundleSizeComment9423"/>

File: change/@fluentui-react-button-aaf5c99d-c063-4562-9ec1-42c216a4b2d7.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-button-aaf5c99d-c063-4562-9ec1-42c216a4b2d7.json:1]: ## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31700&commitId=9c14726e6b2cc13ba33afef3ac2707de9efe6599&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask.normal.chromium.png|3| Changed|

</details>


  <details><summary><b>Card Converged</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged.appearance - High Contrast.normal.chromium.png|0| Added|
|Card Converged.appearance - RTL.normal.chromium.png|0| Added|
|Card Converged.orientation.chromium.png|0| Removed|

</details>


  <details><summary><b>Card Converged - Interactive</b> 9 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged - Interactive.appearance interactive - Outline - High Contrast.click.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Outline - High Contrast.hover.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Outline - High Contrast.normal.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Subtle - Dark Mode.click.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Subtle - Dark Mode.hover.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Subtle - Dark Mode.normal.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Subtle.click.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Subtle.hover.chromium.png|0| Added|
|Card Converged - Interactive.appearance interactive - Subtle.normal.chromium.png|0| Added|

</details>


  <details><summary><b>Card Converged - Selectable</b> 4 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged - Selectable.appearance selectable - Filled - Dark Mode.click.chromium.png|0| Removed|
|Card Converged - Selectable.appearance selectable - Filled - Dark Mode.hover.chromium.png|0| Removed|
|Card Converged - Selectable.appearance selectable - Filled - Dark Mode.normal.chromium.png|0| Removed|
|Card Converged - Selectable.appearance selectable - Filled - Dark Mode.selected.chromium.png|0| Removed|

</details>


  <details><summary><b>Image Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Image Converged.Image Variations Fallback.normal.chromium.png|0| Added|

</details>


  <details><summary><b>SwatchPicker Converged</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|SwatchPicker Converged.spacing.default.chromium.png|0| Added|
|SwatchPicker Converged.Spacing.default.chromium.png|0| Removed|
|SwatchPicker Converged.default - RTL.default.chromium.png|0| Removed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Review by layershifter — CHANGES_REQUESTED** [review]
Let's have a discussion about this on Tech sync first.

For example, I have following concerns:
- this PR adds a bundle size hit
- consumers' overrides for `:hover` selector will stop to work as a new _selector_ will be `@media (any-hover: hover)`

**Comment by spmonahan** [comment]
> Let's have a discussion about this on Tech sync first.
> 
> For example, I have following concerns:
> 
> * this PR adds a bundle size hit
> * consumers' overrides for `:hover` selector will stop to work as a new _selector_ will be `@media (any-hover: hover)`

+1 this. While I really like these newer selectors that, in theory, allow us to better support multiple input modes my past experience has been that they are not uniformly implemented and need lots of testing. As part of further discussion I'd propose we have a device support/testing matrix for features like this.

**Comment by mainframev** [comment]
@layershifter @spmonahan thank you guys! As suggested, let's discuss more in detail on tech sync 👍🏻

**Comment by mainframev** [comment]
Closing in favor of feature request #31858.

**Pull request closed without merging**